### ISSUE TITLE
Instrumentation blocks not saving/restoring correct registers.

### ISSUE THREAD

**Issue by pstach** [dyninst/dyninst#461]
The merged fix for issue #434 (pull req #437) does not resolve the issue of register save/restores on x86-64.

Reproduction details:
1. Modifications are required to the codeCoverage example due to other issues:
   - binary parser library incorrectly reports main executable as shared library
   - skipLibraries needs to contain libInst.C
  A diff is included in the attached zip.
2. Build dyninst off of master, as of today (May 4th)
3. Run "./codeCoverage -sbp /bin/gzip gzip.ins" where gzip is from the ArchLinux gzip-1.9-1 package and has a md5sum of 19eed64fe6cfb8324d1a7a3f9b3dabe3.
4
[repro.zip](https://github.com/dyninst/dyninst/files/1975755/repro.zip)
. run gzip.ins in your debugger of choice.

The outputted binary contains multiple places where rdi is not saved and restored as reported in #434.  Additionally there are a few locations, including the crashing one, where rax is not correctly restored.

From the gzip binary:
objdump -SDrt /bin/gzip
... snip ...
    f436:       49 c1 e5 05             shl    $0x5,%r13 **// start of first example block**
    f43a:       4a 8d 04 29             lea    (%rcx,%r13,1),%rax
    f43e:       8b 10                   mov    (%rax),%edx
    f440:       85 d2                   test   %edx,%edx
    f442:       0f 85 ce 05 00 00       jne    fa16 <stdout@@GLIBC_2.2.5-0x2059e
a>
    f448:       44 89 18                mov    %r11d,(%rax) **// start of second example block**

First example block (offset f436 in original binary):
gdb) x/58i 0x5555559f1930
   0x5555559f1930:      lea    -0xa8(%rsp),%rsp
   0x5555559f1938:      mov    %rax,0x20(%rsp)
   0x5555559f193d:      lea    0xa8(%rsp),%rax
   0x5555559f1945:      and    $0xffffffffffffffe0,%rsp
   0x5555559f1949:      mov    %rax,(%rsp)
   0x5555559f194d:      mov    -0x88(%rax),%rax
   0x5555559f1954:      push   %r10
   0x5555559f1956:      push   %r11
   0x5555559f1958:      push   %r8
   0x5555559f195a:      push   %r9
   0x5555559f195c:      push   %rcx
   0x5555559f195d:      push   %rsi
   0x5555559f195e:      movabs $0xf436,%rax
   0x5555559f1968:      push   %rax
   0x5555559f1969:      push   %rbp
   0x5555559f196a:      mov    %rsp,%rbp
   0x5555559f196d:      lea    -0x200(%rsp),%rsp
   0x5555559f1975:      mov    %rsp,%rax
   0x5555559f1978:      movdqa %xmm0,(%rax)
   0x5555559f197c:      movdqa %xmm1,0x10(%rax)
   0x5555559f1981:      movdqa %xmm2,0x20(%rax)
   0x5555559f1986:      movdqa %xmm3,0x30(%rax)
   0x5555559f198b:      movdqa %xmm4,0x40(%rax)
   0x5555559f1990:      movdqa %xmm5,0x50(%rax)
   0x5555559f1995:      movdqa %xmm6,0x60(%rax)
   0x5555559f199a:      movdqa %xmm7,0x70(%rax)
   0x5555559f199f:      movabs $0x0,%rax
   0x5555559f19a9:      callq  *-0x18bc47(%rip)        # 0x555555865d68
   0x5555559f19af:      test   %rax,%rax
   0x5555559f19b2:      je     0x5555559f19e2
   0x5555559f19b8:      movabs $0x1023,%rdi
   0x5555559f19c2:      movabs $0x0,%rax
   0x5555559f19cc:      callq  *-0x18bc52(%rip)        # 0x555555865d80
   0x5555559f19d2:      movabs $0x0,%rax
   0x5555559f19dc:      callq  *-0x18bc6a(%rip)        # 0x555555865d78
   0x5555559f19e2:      mov    %rsp,%rax
   0x5555559f19e5:      movdqa (%rax),%xmm0
   0x5555559f19e9:      movdqa 0x10(%rax),%xmm1
   0x5555559f19ee:      movdqa 0x20(%rax),%xmm2
   0x5555559f19f3:      movdqa 0x30(%rax),%xmm3
   0x5555559f19f8:      movdqa 0x40(%rax),%xmm4
   0x5555559f19fd:      movdqa 0x50(%rax),%xmm5
   0x5555559f1a02:      movdqa 0x60(%rax),%xmm6
   0x5555559f1a07:      movdqa 0x70(%rax),%xmm7
   0x5555559f1a0c:      leaveq 
   0x5555559f1a0d:      pop    %rax
   0x5555559f1a0e:      pop    %rsi
   0x5555559f1a0f:       pop    %rcx
   0x5555559f1a10:      pop    %r9
   0x5555559f1a12:      pop    %r8
   0x5555559f1a14:      pop    %r11
   0x5555559f1a16:      pop    %r10
   0x5555559f1a18:      mov    (%rsp),%rsp
   0x5555559f1a1c:      shl    $0x5,%r13
   0x5555559f1a20:      lea    (%rcx,%r13,1),%rax
   0x5555559f1a24:      mov    (%rax),%edx
   0x5555559f1a26:      test   %edx,%edx
   0x5555559f1a28:      jne    0x5555559f7a8d

One can see it uses rdi but does not save or restore it.

Second example block (offset f448 in original binary)
(gdb) x/56i 0x5555559f1a2e
   0x5555559f1a2e:      lea    -0xa8(%rsp),%rsp
   0x5555559f1a36:      mov    %rax,0x20(%rsp)
   0x5555559f1a3b:      lea    0xa8(%rsp),%rax
   0x5555559f1a43:      and    $0xffffffffffffffe0,%rsp
   0x5555559f1a47:      mov    %rax,(%rsp)
   0x5555559f1a4b:      mov    -0x88(%rax),%rax
   0x5555559f1a52:      push   %r10
   0x5555559f1a54:      push   %r11
   0x5555559f1a56:      push   %r8
   0x5555559f1a58:      push   %r9
   0x5555559f1a5a:      push   %rcx
   0x5555559f1a5b:      push   %rdx
   0x5555559f1a5c:      push   %rsi
   0x5555559f1a5d:      movabs $0xf448,%rax
   0x5555559f1a67:      push   %rax
   0x5555559f1a68:      push   %rbp
   0x5555559f1a69:      mov    %rsp,%rbp
   0x5555559f1a6c:      lea    -0x218(%rsp),%rsp
   0x5555559f1a74:      mov    %rsp,%rax
   0x5555559f1a77:      movdqa %xmm0,(%rax)
   0x5555559f1a7b:      movdqa %xmm1,0x10(%rax)
   0x5555559f1a80:      movdqa %xmm2,0x20(%rax)
   0x5555559f1a85:      movdqa %xmm3,0x30(%rax)
   0x5555559f1a8a:      movdqa %xmm4,0x40(%rax)
   0x5555559f1a8f:       movdqa %xmm5,0x50(%rax)
   0x5555559f1a94:      movdqa %xmm6,0x60(%rax)
   0x5555559f1a99:      movdqa %xmm7,0x70(%rax)
   0x5555559f1a9e:      movabs $0x0,%rax
   0x5555559f1aa8:      callq  *-0x18bd46(%rip)        # 0x555555865d68
   0x5555559f1aae:      test   %rax,%rax
   0x5555559f1ab1:      je     0x5555559f1ae1
   0x5555559f1ab7:      movabs $0x1024,%rdi
   0x5555559f1ac1:      movabs $0x0,%rax
   0x5555559f1acb:      callq  *-0x18bd51(%rip)        # 0x555555865d80
   0x5555559f1ad1:      movabs $0x0,%rax
   0x5555559f1adb:      callq  *-0x18bd69(%rip)        # 0x555555865d78
   0x5555559f1ae1:      mov    %rsp,%rax
   0x5555559f1ae4:      movdqa (%rax),%xmm0
   0x5555559f1ae8:      movdqa 0x10(%rax),%xmm1
   0x5555559f1aed:      movdqa 0x20(%rax),%xmm2
   0x5555559f1af2:       movdqa 0x30(%rax),%xmm3
   0x5555559f1af7:       movdqa 0x40(%rax),%xmm4
   0x5555559f1afc:       movdqa 0x50(%rax),%xmm5
   0x5555559f1b01:      movdqa 0x60(%rax),%xmm6
   0x5555559f1b06:      movdqa 0x70(%rax),%xmm7
   0x5555559f1b0b:      leaveq 
   0x5555559f1b0c:      pop    %rax
   0x5555559f1b0d:      pop    %rsi
   0x5555559f1b0e:      pop    %rdx
   0x5555559f1b0f:       pop    %rcx
   0x5555559f1b10:      pop    %r9
   0x5555559f1b12:      pop    %r8
   0x5555559f1b14:      pop    %r11
   0x5555559f1b16:      pop    %r10
   0x5555559f1b18:      mov    (%rsp),%rsp
   0x5555559f1b1c:      mov    %r11d,(%rax)

One can see that it uses but does not save or restore rdi.  Additionally the value of rax gets corrupted and restores to 0xf448, which triggers a segfault on 0x5555559f1b1c.

Attached is a copy of the diff to codeCoverage, gzip.orig, from the ArchLinux package, and gzip.ins the instrumented copy.

**Comment by pstach** [comment]
Quick update, in the liveness debug output, it reports that the instruction at f448 is "mov %r11d,(%r8d)" not "mov %r11d,(%rax)" as in the source binary.

**Comment by pstach** [comment]
The REX bits in prefixes are handled incorrectly in a number of cases.
Examples (in intel syntax):
        cmp byte [rax],0
        mov [r14],rax
        mov [rax],r11d

**Comment by mxz297** [comment]
Thanks for your detailed information. It seems like you are seeing two different issues: 1. rdi & rax are not correctly saved; 2. multiple instructions are wrongly decoded. 

Is the first issue caused by the second one or independent from the second one?

**Comment by pstach** [comment]
Correct, there are two issues here.

1. Instrumentation register dependencies are not correctly tracked.  This is likely in dataflowAPI, or at least that is where I am currently looking.  That is after the decodes are corrected. 
2. The REX prefix byte is incorrectly handled for memory references.  The size and/or register number is computed using the wrong bits from the REX byte.   I'm looking into it, however the disassembler in libcommon is structured a good bit different than any other x86 disassembler I've seen.

**Comment by pstach** [comment]
Found the source of issue 2.  The cases in https://github.com/dyninst/dyninst/blob/master/common/src/arch-x86.C#L9713 don't process REX byte unless macadr is non-null.

**Comment by mxz297** [comment]
It looks to me that we are wrongly applying the REX byte. Based on the "apply_rex_bit" function in common/h/arch-x86.h, if Dyninst reports "(%r8d)", rather than "(%rax)", it means Dyninst thought that the REX.B bit is 1. However, it should be zero. 

Can you confirm that for that instruction, we indeed have REX.B bit set to 1?

By the way, I am busy with other issues. I appreciate your helps of digging into the code base and investigating these issues.

**Comment by pstach** [comment]
For mov %r11d,(%rax), the REX byte is 0x44, so REX.R is the only bit set.  I've isolated a way to test with just InstructionDecoder, however there appear to be a number of problems due to postprocessing the REX byte.

I'm working on a patch that processes REX immediately when doing operand decoding, however this impacts other areas of the source tree.  In general it seems like the cleanest way to resolve.  I also found a few other issues in the x86 disassembler regarding addressing size that this should fix as well.
